src/mlpack/methods/ne/neat.hpp

```diff
@@ -0,0 +1,909 @@
+      ssize_t speciesSize = population.aSpecies[i].aGenomes.size();
+      if (speciesSize > 0) {
+        Genome bestGenome = population.aSpecies[i].aGenomes[0];
+        population.aSpecies[i].aGenomes.clear();
+        population.aSpecies[i].aGenomes.push_back(bestGenome);
+      }
+    }
+    RemoveEmptySpecies(population);
+  }
+
+  // Mutate child by combining different mutations.
+  // NOTICE: how we organize different mutations is kind of flexible.
+  void Mutate(Genome& genome) {
+    // Mutate weights.
+    MutateWeight(genome, aMutateWeightProb, aPerturbWeightProb, aMutateWeightSize);
+
+    // Mutate add link.
+    double p = aMutateAddLinkProb;
+    while (p > 0) {  // so p can be bigger than 1 and mutate can happen multiple times.
+      if (mlpack::math::Random() < p) {
+        MutateAddLink(genome, aMutateAddLinkProb, aMutateAddRecurrentLinkProb, aMutateAddLoopLinkProb);
+      }
+      --p;
+    }
+
+    // Mutate add neuron.
+    p = aMutateAddNeuronProb;
+    while (p > 0) {
+      if (mlpack::math::Random() < p) {
+        MutateAddNeuron(genome, aMutateAddNeuronProb);
+      }
+      --p;
+    }
+
+    // Mutate enabled node to disabled.
+    p = aMutateEnabledProb;
+    while (p > 0) {
+      if (mlpack::math::Random() < p) {
+        MutateEnableDisable(genome, true, aMutateEnabledProb);
+      }
+      --p;
+    }
+
+    // Mutate disabled node to enabled.
+    p = aMutateDisabledProb;
+    while (p > 0) {
+      if (mlpack::math::Random() < p) {
+        MutateEnableDisable(genome, false, aMutateDisabledProb);
+      }
+      --p;
+    }
+  }
+
+  // Breed child for a species.
+  // Crossover to born a child, or copy a child, then mutate it.
+  // NOTICE: can have different ways to breed a child.
+  bool BreedChild(Species& species, Genome& childGenome, double crossoverProb) {
+    double p = mlpack::math::Random();
+    ssize_t speciesSize = species.aGenomes.size();
+
+    if (speciesSize == 0)
+      return false;
+
+    if (p < crossoverProb) {
+      ssize_t idx1 = mlpack::math::RandInt(0, speciesSize);
+      ssize_t idx2 = mlpack::math::RandInt(0, speciesSize);
+      if (idx1 != idx2) {
+        Crossover(species.aGenomes[idx1], species.aGenomes[idx2], childGenome);
+      } else {
+        return false;
+      }
+    } else {
+      ssize_t idx = mlpack::math::RandInt(0, speciesSize);
+      childGenome = species.aGenomes[idx];
+    }
+
+    Mutate(childGenome);
+
+    return true;
+  }
+
+
+  // Initialize population.
+  void InitPopulation() {
+    aPopulation = Population(aSeedGenome, aPopulationSize);
+  }
+
+  // Reproduce next generation of population.
+  void Reproduce() {
+    // keep best genome.
+    Genome lastBestGenome = aPopulation.BestGenome();
+
+    // Remove stale species.
+    if (aPopulation.aSpecies.size() > 10) {
+      RemoveStaleSpecies(aPopulation);
+    }
+
+    // Remove weak genomes in each species.
+    CullSpecies(aPopulation, aCullSpeciesPercentage);
```

Comment:
Thanks for the information, maybe it's a good idea, but it sounds counterproductive to me. I think it should work without, since RemoveStaleSpecies should remove old species anyway.